File: skills/pretext/scripts/check_pretext_eval_coverage.py
#!/usr/bin/env python3
"""Check whether the formal eval suite covers all supported goals and surfaces."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from select_pretext_api import SUPPORTED_GOALS, SUPPORTED_SURFACES


def main() -> int:
    repo_root = Path(__file__).resolve().parents[3]
    evals_path = repo_root / "skills" / "pretext" / "evals" / "evals.json"
    coverage_path = repo_root / "skills" / "pretext" / "evals" / "coverage.json"

    evals_data = json.loads(evals_path.read_text(encoding="utf-8"))
    coverage_data = json.loads(coverage_path.read_text(encoding="utf-8"))

    eval_ids = {item["id"] for item in evals_data["evals"]}
    missing_eval_ids: list[int] = []
    for goal, ids in coverage_data["goals"].items():
        for eval_id in ids:
            if eval_id not in eval_ids and eval_id not in missing_eval_ids:
                missing_eval_ids.append(eval_id)
    for surface, ids in coverage_data["surfaces"].items():
        for eval_id in ids:
            if eval_id not in eval_ids and eval_id not in missing_eval_ids:
                missing_eval_ids.append(eval_id)

    missing_goals = [goal for goal in SUPPORTED_GOALS if goal not in coverage_data["goals"]]
    missing_surfaces = [surface for surface in SUPPORTED_SURFACES if surface != "generic" and surface not in coverage_data["surfaces"]]
    empty_goals = [goal for goal in SUPPORTED_GOALS if not coverage_data["goals"].get(goal)]
    empty_surfaces = [
        surface
        for surface in SUPPORTED_SURFACES
        if surface != "generic" and not coverage_data["surfaces"].get(surface)
    ]

    if not missing_goals and not missing_surfaces and not empty_goals and not empty_surfaces and not missing_eval_ids:
        print("Eval coverage is in sync with supported goals and surfaces.")
        return 0

    if missing_goals:
        print("Missing goals in coverage.json:")
        for goal in missing_goals:
            print(f"- {goal}")
    if empty_goals:
        print("Goals without mapped evals:")
        for goal in empty_goals:
            print(f"- {goal}")
    if missing_surfaces:
        print("Missing surfaces in coverage.json:")
        for surface in missing_surfaces:
            print(f"- {surface}")
    if empty_surfaces:
        print("Surfaces without mapped evals:")
        for surface in empty_surfaces:
            print(f"- {surface}")
    if missing_eval_ids:
        print("Coverage references unknown eval IDs:")
        for eval_id in sorted(missing_eval_ids):
            print(f"- {eval_id}")
    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/pretext/scripts/check-pretext-eval-coverage.ts
#!/usr/bin/env bun
/**
 * Check whether the formal eval suite covers all supported goals, surfaces,
 * and validation areas.
 *
 * Usage:
 *   bun run skills/pretext/scripts/check-pretext-eval-coverage.ts
 *
 * Exit code 0 if in sync, 1 if mismatches found.
 */

import { resolve, join } from "node:path";

// ── Inline catalog: the 13 validation areas (matching select-pretext-validation-from-git) ──

const VALIDATION_AREAS = [
  "analysis",
  "measurement",
  "line-break",
  "layout-api",
  "bidi",
  "benchmark-harness",
  "accuracy-harness",
  "probe-surface",
  "corpus-tooling",
  "gatsby-surface",
  "package-workflow",
  "demo-site",
  "reporting-tooling",
] as const;

// ── Types ──────────────────────────────────────────────────────────────────

interface EvalsData {
  evals: { id: number; [key: string]: unknown }[];
}

interface CoverageData {
  goals: Record<string, number[]>;
  surfaces: Record<string, number[]>;
  validation_areas?: Record<string, number[]>;
  [key: string]: unknown;
}

// ── Helpers ────────────────────────────────────────────────────────────────

/** Collect all eval IDs referenced in a category map, checking for unknown IDs. */
function collectMissingIds(
  categoryMap: Record<string, number[]> | undefined,
  validIds: Set<number>,
  missingIds: number[],
): void {
  if (!categoryMap) return;
  for (const ids of Object.values(categoryMap)) {
    for (const evalId of ids) {
      if (!validIds.has(evalId) && !missingIds.includes(evalId)) {
        missingIds.push(evalId);
      }
    }
  }
}

function reportList(header: string, items: string[]): void {
  if (items.length === 0) return;
  console.log(header);
  for (const item of items) {
    console.log(`- ${item}`);
  }
}

// ── Main ───────────────────────────────────────────────────────────────────

async function main(): Promise<number> {
  // import.meta.dir → …/skills/pretext/scripts  →  3 parents up = repo root
  const repoRoot = resolve(import.meta.dir, "..", "..", "..");

  const evalsPath = join(repoRoot, "skills", "pretext", "evals", "evals.json");
  const coveragePath = join(repoRoot, "skills", "pretext", "evals", "coverage.json");

  const evalsData: EvalsData = await Bun.file(evalsPath).json();
  const coverageData: CoverageData = await Bun.file(coveragePath).json();

  const evalIds = new Set(evalsData.evals.map((e) => e.id));

  // ── Check for unknown eval IDs in coverage ──
  const missingEvalIds: number[] = [];
  collectMissingIds(coverageData.goals, evalIds, missingEvalIds);
  collectMissingIds(coverageData.surfaces, evalIds, missingEvalIds);
  collectMissingIds(coverageData.validation_areas, evalIds, missingEvalIds);

  // ── Check for missing / empty goals ──
  const goalKeys = Object.keys(coverageData.goals);
  // Goals are user-defined; we just check that the coverage file has them and they're non-empty.
  const emptyGoals = goalKeys.filter((g) => !coverageData.goals[g]?.length);

  // ── Check for missing / empty surfaces ──
  const surfaceKeys = Object.keys(coverageData.surfaces);
  const emptySurfaces = surfaceKeys.filter((s) => !coverageData.surfaces[s]?.length);

  // ── Check for missing / empty validation areas ──
  const coveredValidation = coverageData.validation_areas ?? {};
  const missingValidationAreas = VALIDATION_AREAS.filter(
    (area) => !(area in coveredValidation),
  );
  const emptyValidationAreas = VALIDATION_AREAS.filter(
    (area) => area in coveredValidation && !coveredValidation[area]?.length,
  );

  // ── Verdict ──
  const allClear =
    emptyGoals.length === 0 &&
    emptySurfaces.length === 0 &&
    missingValidationAreas.length === 0 &&
    emptyValidationAreas.length === 0 &&
    missingEvalIds.length === 0;

  if (allClear) {
    console.log(
      "Eval coverage is in sync with goals, surfaces, and validation areas.",
    );
    return 0;
  }

  reportList("Goals without mapped evals:", emptyGoals);
  reportList("Surfaces without mapped evals:", emptySurfaces);
  reportList("Missing validation areas in coverage.json:", missingValidationAreas);
  reportList("Validation areas without mapped evals:", emptyValidationAreas);

  if (missingEvalIds.length > 0) {
    console.log("Coverage references unknown eval IDs:");
    for (const id of missingEvalIds.sort((a, b) => a - b)) {
      console.log(`- ${id}`);
    }
  }

  return 1;
}

process.exit(await main());
